src/reducers/computeRenderedRows.js

```diff
@@ -239,3 +245,97 @@ function computeRenderedRowOffsets(state, rowRange) {
   state.rowOffsets = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+
+/**
+ * Walk the rows to render and compute the height offsets and
+ * positions in the row buffer.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdx: number,
+ *   endViewportIdx: number,
+ *   firstBufferIdx: number,
+ *   firstViewportIdx: number,
+ * }} rowRange
+ * @private
+ */
+function computeRenderedRowOffsetsInViewport(state, rowRange) {
+  const { rowBufferSet, rowOffsetIntervalTree, storedHeights } = state;
+  const {
+    endBufferIdx,
+    endViewportIdx,
+    firstBufferIdx,
+    firstViewportIdx,
+  } = rowRange;
+
+  const renderedRowsCount = endBufferIdx - firstBufferIdx;
+  if (renderedRowsCount === 0) {
+    state.rowOffsets = {};
+    state.rows = [];
+    return;
+  }
+
+  // output for this function
+  const rows = state.rows.slice(); // state.rows
```

Comment:
Document that you're doing this to clone the array

I'm also not convinced this is the best way to ensure the rows list stays the same length.  It's a bit obscure

Reply:
`rows` is strictly tied to the buffer. Hence buffer range decides the `rows.length`, and usually this is always the same length as the buffer. 
If the buffer range increases, then the `rows.length` length also goes up.

However, when the buffer range reduces from the previous one, the rows list doesn't decrease in length. 
But this is actually a performance optimization as it hinders unmounts. We could additionally have simple logic to collapse the range back up at the time of scroll end. (This was an optimization we offlined about but hadn't made an issue out of it)

Anyway, we're always calculating `rowOffsets` for every row in `rows`, so this shouldn't lead to a problem?

Reply:
My biggest concern with this PR is that what you've explained here is obscure given how the code is written and has a high risk of breaking under future maintenance.  We could write some unit tests which ensure it won't be.

But I think a better approach would be to write code which implements this directly.  If our goal is to ensure we don't unmount rows while scrolling.  Could pass to the component:
1) rows - an ordered list of row indexes to render.  The code here should ensure a row index in the viewport is mapped to the same ordered position in the array as it was on the last render.
2) rowOffsets - a map from row index to the offset positioning for each row in the viewport.
3) countOfRowsToRender - this should be the max number of rows within the viewport during this segment of scrolling

Then the renderer can do:
```
for (let pos = 0; pos < countOfRowsToRender; pos++) {
  if (rows[pos] !== undefined) {
      // render row as normal at offset
  } else {
      // render dummy row w/ display: none styling
  }
}
```

I think that should be possible and would make a much cleaner separation of concerns.

While not scrolling things can remain similar where the rows & offsets also include the buffer (along w/ the viewport) and the countOfRowsToRender can be viewport+buffer.

Does that sound reasonable?  I think a lot of my concern here is with trying to address the padding off rows to avoid unmounts in this code.  It's too obscure to read the code an know that's important to do.